File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cpp
class KthLargest {
private:
    priority_queue<int, vector<int>, greater<int>> pq{};
    int k;

public:
    KthLargest(int k, vector<int>& nums) : k{k} {
        int n = nums.size();
        int m = min(k, n);
        int i = 0;

        for (; i < m; ++i) {
            pq.push(nums[i]);
        }

        for (; i < n; ++i) {
            add(nums[i]);
        }
    }
    
    int add(int val) {
        if (pq.size() == k) {
            int top = pq.top();

            if (val <= top) {
                return top;
            }

            pq.pop();
        }
        
        pq.push(val);
        return pq.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

File: 0703-kth-largest-element-in-a-stream/0703-kth-largest-element-in-a-stream.cs
public class KthLargest {

    private PriorityQueue<int, int> pq;
    private int cap;

    public KthLargest(int k, int[] nums) {
        cap = k;
        pq = new PriorityQueue<int, int>(k);

        foreach (int num in nums) {
            Add(num);
        }
    }
    
    public int Add(int val) {
        if (pq.Count == cap) {
            int top = pq.Peek();

            if (val <= top) {
                return top;
            }

            pq.Dequeue();
        }

        pq.Enqueue(val, val);
        return pq.Peek();
    }
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.Add(val);
 */